File: Solutions/227_Basic_Calculator_II/227_Basic_Calculator_II.js
//https://leetcode.com/problems/basic-calculator-ii/discuss/597084/Javascript-stack-O(N)
/**
 * @param {string} s
 * @return {number}
 */
var calculate = function(s) {
  let stack = [];
  let num = '';
  let sign = null
  // we loop till the full length of the array to account for last sign
  for(let i = 0; i <= s.length; i++){    
    const curr = s[i];
    //handle space
    if(curr === ' ') continue;
    //if char is a number
    if(!isNaN(curr)) num+=curr;
    //if we have a  sign + - / *
    if(isNaN(curr)){
      num = Number(num)
      switch(sign){
        case '+':
        case null:
          //we push the initial number into the stack
          stack.push(num)
          break;
        case '-':
          //we push any values after the subtraction sign as negative
          stack.push(-num)
          break; 
        case '*':
          //we pop the stack then multiply and push back
          stack.push(stack.pop()*num)
          break;
        case '/':
          //we pop the stack then devide and push back
          stack.push(parseInt(stack.pop()/num, 10))
          break;           
      }
      // sign becomes current sign
      sign = curr;
      // we reset num
      num = '';
    }
  }
  //we reduce the array adding positive and negative numbers
  return stack.reduce((a,b)=>{
    return a+b
  },0)
};

//Third fastest
var calculate = function(s) {
    const stack = []
    let num = 0
    let operator = '+'
    let i = 0
    
    while (i < s.length) {
        const char = s[i]
                
        if (isDigit(char)) {
            let curr = 0
            while (i < s.length && isDigit(s[i])) {
                curr *= 10
                curr += +s[i]
                i++
            }
            
            if (operator === '+') {
                stack.push(curr)
            } else if (operator === '-') {
                stack.push(-curr)
            } else if (operator === '/') {
                stack.push(Math.trunc(stack.pop() / curr))
            } else if (operator === '*') {
                stack.push(stack.pop() * curr)
            }
            
            continue
        }
        
        if (char !== ' ') operator = char
        i++
        continue
    }
    
    let result = 0
    while (stack.length) {
        result += stack.pop()
    }
    return result
};

const isDigit = char => char >= '0' && char <= '9'

//Second fastest
var calculate = function(s) {
    s = s.replace(/\s+/g,'');
    var stack = []
    var num = 0
    var sign = '+';
        
    for(var i = 0;i < s.length;i++){
        let c = s[i];
        // console.log(stack);
        if(c >=0 && c <= 9){
            num = num * 10 + c * 1;
        }
        if(['+','-','*','/'].indexOf(c) > -1 || i == s.length - 1){
            if(sign == '+'){
                stack.push(num);
            }else if(sign == '-'){
                stack.push(-num);
            }else if(sign == '*'){
                stack.push(stack.pop() * num)
            }else if(sign == '/'){
                stack.push((stack.pop() / num) >> 0);
            }
            sign = c;
            num = 0;
        }
    }
    
    return stack.reduce((accu,cur) => accu + cur,0);
};

//Fastest
var calculate = function(s) {
    if (s.length === 0) {
       return 0;
   }
   
   s = s.replace(/\s+/g, "");

   let op = '+';
   let stack = [];
   let n = "";
   
   for (let i=0; i <= s.length; i++) {
       let c = s[i]
        
       if(c >= '0' && c <= '9') {
           n += c;
       } else {
           if (op === '+') {
               stack.push(parseInt(n));
           } else if (op === '-') {
               stack.push(parseInt(-n))
           } else if (op === '*') {
               stack.push(stack.pop() * parseInt(n));
           } else if (op === '/') {
               stack.push(Math.trunc(stack.pop() / parseInt(n)));
           }
       
           op = c;
           n = "";  
       }
   }
   // return the sum of the stack.
   return stack.reduce((a,b) => a+b, 0); 
};

File: Solutions/227_Basic_Calculator_II/227_Basic_Calculator_II.rb
#Third fastest
# 给一个有乘除法，加减的算式，给出计算结果

# 思路
# 1. we can treat different characters differently
# 2. for number, we can just add to the number we have so far
# 3. and for operators, we can get the result based on last operator and then push the result into the stack
# 4. finally we can sum the result from the stack and that should be the answer

# 如果有括号
# 5. for open bracket, we can do recursion from this index
# 6. for close bracket, we should return the result to the level above


class Calculator
    @@numbers = (0..9).map { |num| num.to_s }
    def initialize(expression)
      @expression = (expression + '+').chars
    end
  
    def evaluate
      number = 0
      stack = []
      operator = '+'
      until @expression.empty?
        char = @expression.shift
        next if char == ' '
  
        # if it;s just a regular number
        if @@numbers.include?(char)
          number = number * 10 + char.to_i
          next
        end
  
        # =begin TODO recursion: one level down
        # if char == '('
        #   number = evaluate() # TODO
        #   next
        # end
        # =end TODO
  
        # conclude the window if not a digit
        if operator == '+'
          stack.push(number)
        elsif operator == '-'
          stack.push(-1 * number)
        elsif operator == '*'
          stack.push(stack.pop * number)
        elsif operator == '/'
          divisor = stack.pop
          stack.push(
            (divisor >= 0 ? 1 : -1) * (divisor.abs / number)
          )
        end
        operator = char
        number = 0
  
        # break and return back to the level above
        # break if char == ')' # TODO
      end
      stack.sum
    end
  end
  
  def calculate(s)
    Calculator.new(s).evaluate
  end

#Second fastest
def evaluate(left, right, op)
    case op
        when '+'
            return left + right
        when '-'
            return left - right
        when '*'
            return left * right
        when '/'
            return left / right
        else
            return 0
    end
end

def calculate(s)
    digits = '0123456789'.chars
    operations = '+-*/'.chars
    result = 0
    nums = []
    ops = []
    num = 0
    s.chars().each do |c|
        next if c == ' '
        if digits.include?(c)
            num = num * 10 + c.to_i
        else operations.include?(c)
            nums << num
            num = 0
            if ops.length > 0
                if ops[-1] == '*' || ops[-1] == '/'
                    right = nums.pop rescue 1
                    left = nums.pop rescue 1
                    op = ops.pop
                    nums << evaluate(left, right, op)
                end
            end
            ops << c
        end
    end

    nums << num
    if ops.length > 0
        if ops[-1] == '*' || ops[-1] == '/'
            right = nums.pop rescue 1
            left = nums.pop rescue 1
            op = ops.pop
            nums << evaluate(left, right, op)
        end
    end
    
    index = 0
    val = nums[index]
    ops.each do |op|
        index += 1
        val = evaluate(val, nums[index], op)
    end
    
    val
end

#Fastest
def calculate(s)
    num = 0
    res = 0
    prod = 1
    inverse = false
    sign = 1
    s.chars.each do |c|
        next if c == ' '
        if c == '+'
            if inverse
                prod /= num
            else
                prod *= num
            end
            res += prod * sign
            num = 0
            sign = 1
            prod = 1
            inverse = false
        elsif c == '-'
            if inverse
                prod /= num
            else
                prod *= num
            end
            res += prod * sign
            num = 0
            sign = -1
            prod = 1
            inverse = false
        elsif c == '*'
            if inverse
                prod /= num
            else
                prod *= num
            end
            inverse = false
            num = 0
        elsif c == '/'
            if inverse
                prod /= num
            else
                prod *= num
            end
            num = 0
            inverse = true
        else
            num = num * 10 + c.to_i
        end
    end
    if inverse
        prod /= num
    else
        prod *= num
    end
    res += prod * sign
    res
end